client/client-lib/src/lib.rs

```diff
@@ -698,9 +733,17 @@ impl Client<UserClientConfig> {
             )
             .json(&contract_id)
             .send();
-        fedimint_api::task::timeout(Duration::from_secs(15), future)
-            .await
-            .map_err(|_| ClientError::OutgoingPaymentTimeout)??;
+        let gateway_future = async {
+            fedimint_api::task::timeout(Duration::from_secs(15), future)
+                .await
+                .map_err(|_| ClientError::OutgoingPaymentTimeout)??;
+            Result::<()>::Ok(())
+        };
+
+        if gateway_future.await.is_err() {
+            self.try_refund_outgoing_contract(contract_id, rng).await?;
+        }
```

Comment:
This doesn't work.

It only calls `try_refund_outgoing_contract` when `/pay_invoice` call times out. But if it does, then it's almost guaranteed that the gateway hasn't aborted the outgoing payment and had federation accept it. So this will fail.

The following makes this feature work in Fluttermint:

```rust
let result = fedimint_api::task::timeout(Duration::from_secs(15), future)
    .await
    .map_err(|_| ClientError::OutgoingPaymentTimeout)?
    .map_err(|e| ClientError::HttpError(e));

if let Ok(response) = result {
    if !response.status().is_success() {
	// wait for gateway cancellation to be accepted by federation
	fedimint_api::task::sleep(std::time::Duration::from_secs(1)).await;
	self.try_refund_outgoing_contract(contract_id, rng).await?;
    }
};

```

Also, in the case that `try_refund_outgoing_contract` is called, we should return `Err(ClientError::OutgoingPaymentAborted(outpoint))` or something. Our outgoing contract wasn't fully executed. A client would want to show an error here. With that outpoint the client could also call `fetch_coins(outpoint)` to recover ecash tokens.

Reply:
Will apply this change. We still want to keep track of the payment in case of timeout (or increase the timeout to like 2min, so the GW will fail before nearly always, cln has a 1min timeout iirc).

Reply:
As mentioned earlier, we can have clients like Fluttermint poll the database periodically to see if pending outgoing payments have been cancelled and attempt to sweep the funds back.